rules/collapsible_heading.js

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2024 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+'use strict';
+
+/**
+ * @module modifiers/link_attributes
+ */
+
+const safeObject = require('../util/safe_object');
+
+/**
+ * @typedef {Object} CollapsibleHeadingOptions
+ * @property {string[]} [levels] List of headings to transform.
+ * @property {boolean} [open=true] Whether to collapse the content by default.
+ */
+
+/**
+ * Add support for mentioning users, using an `@` symbol. Wraps the mention in a link to the user.
+ *
+ * By default, any characters that are not a space or newline after an `@` symbol will be treated as a mention.
+ *
+ * @example
+ * Hello @test
+ *
+ * <p>Hello <a href="/users/test">@test</a></p>
+ *
+ * @type {import('markdown-it').PluginWithOptions<CollapsibleHeadingOptions>}
+ */
+module.exports = (md, options) => {
+    // Get the correct options
+    const optsObj = safeObject(options);
+
+    /**
+     * Rule for inserting details around headings.
+     *
+     * @type {import('markdown-it').RuleCore}
+     * @private
+     */
+    const collapsibleHeading = state => {
+        const collapsed = [];
+        state.tokens = state.tokens.reduce((newTokens, token) => {
+            if (token.type === 'heading_open') {
+                const headingLevel = Number(token.tag.replace('h', ''));
+                // Close all hanging collapsible sections.
+                while (collapsed[collapsed.length - 1] >= headingLevel) {
+                    // Close the previous detail element
+                    const closeDetail = new state.Token('details', 'details', -1);
+                    closeDetail.block = true;
+                    newTokens.push(closeDetail);
+
+                    // Remove the closed collapsible from the list.
+                    collapsed.pop();
+                }
+
+                if (Array.isArray(optsObj.levels) && optsObj.levels.includes(token.tag)) {
+                    // Create the outer details element
+                    const openDetails = new state.Token('details', 'details', 1);
+                    openDetails.block = true;
+                    openDetails.attrSet('class', 'collapsible');
```

Comment:
Allow this to be customized via an opt, like we do for other plugins?

Reply:
Changed.